### PR TITLE
Allow validators to block and kick their nominator set.

### PR DESCRIPTION
**Pull request by gavofyork**
This allows validators to curate their nominator set if desired. Default behaviour remains unchanged.

Two main changes:
- `ValidatorPrefs` now contains an additional `bool` field: `blocked`. This is `false` by default (which give current behaviour), but if set to `true`, then nobody who is not already nominating the validator may nominate it.
- There is an additional dispatchable: `kick`. This is called by a validator (controller account) and takes a list of nominator (stash accounts); afterwards those accounts are no longer nominating the validator.

You'll typically use it by:
1. Create a validator
2. Nominate the validator with whatever accounts you want
3. Switch validator prefs to *block* any further nominations
4. If there are any nominations that are undesirable, then `kick` them

CC @jacogr @kianenigma @shawntabrizi 

Polkadot companion: paritytech/polkadot#2290

### CODE REVIEW

**Review by kianenigma — APPROVED** [review]
LGTM, one potential improvement is that the migration is a one-time use and while we haven't made anything standard yet, it is better to make standalone ones that don't rely on a particular `T: Config`, [as such](https://github.com/paritytech/substrate/pull/7040/files#diff-3acd0cb73093b0cc01acfd0dc5646b31a309e57526f18d67d7e9d8708c325dbb). Then the migration can live forever and re-used much nicer. But it is a bit of hassle to set up and since not everyone is doing it, I am not sure how valuable it is

File: frame/staking/src/lib.rs [hunk 1]
```diff
@@ -1675,9 +1719,17 @@ decl_module! {
...
 			ensure!(!targets.is_empty(), Error::<T>::EmptyTargets);
+			ensure!(targets.len() <= MAX_NOMINATIONS, Error::<T>::TooManyTargets);
+
+			let old = Nominators::<T>::get(stash).map_or_else(Vec::new, |x| x.targets);
+
 			let targets = targets.into_iter()
-				.take(MAX_NOMINATIONS)
-				.map(|t| T::Lookup::lookup(t))
+				.map(|t| T::Lookup::lookup(t).map_err(DispatchError::from))
+				.map(|n| n.and_then(|n| if old.contains(&n) || !Validators::<T>::get(&n).blocked {
```
Comment (kianenigma) [inline frame/staking/src/lib.rs:1728]: I think this reads nicer if the `if-else` body is indented.

**Comment by gavofyork** [comment]
> LGTM, one potential improvement is that the migration is a one-time use and while we haven't made anything standard yet, it is better to make standalone ones that don't rely on a particular `T: Config`, [as such](https://github.com/paritytech/substrate/pull/7040/files#diff-3acd0cb73093b0cc01acfd0dc5646b31a309e57526f18d67d7e9d8708c325dbb). Then the migration can live forever and re-used much nicer. But it is a bit of hassle to set up and since not everyone is doing it, I am not sure how valuable it is

The huge issue with this is that it quickly becomes painful to support multiple migrations, each with their own datastructures, when they change on top of each other. Our storage `transform` API is set up to migrate from a generic data structure into the current (as defined in the storage item) data structure. Migrating to anything other than the current data structure is far more painful. Realistically, migrations won't "live forever" in any meaningful way, unless you want to dramatically increase the burden of the author by requiring them use custom storage transforms/iteration code than can handle going into a type which is not the current storage item's type. This, I think, will lend itself to tedious and error-prone code.

File: frame/staking/src/lib.rs [hunk 2]
```diff
@@ -2168,6 +2220,42 @@ decl_module! {
...
+		///
+		/// - `who`: A list of nominator stash accounts who are nominating this validator which
+		///   should no longer be nominating this validator.
+		///
+		/// Note: Making this call only makes sense if you first set the validator preferences to
+		/// block any further nominations.
+		#[weight = T::WeightInfo::kick(who.len() as u32)]
+		pub fn kick(origin, who: Vec<<T::Lookup as StaticLookup>::Source>) -> DispatchResult {
+			ensure!(Self::era_election_status().is_closed(), Error::<T>::CallNotAllowed);
+			let controller = ensure_signed(origin)?;
```
Comment (shawntabrizi) [inline frame/staking/src/lib.rs:2240]:
```suggestion
			let controller = ensure_signed(origin)?;
			ensure!(Self::era_election_status().is_closed(), Error::<T>::CallNotAllowed);
```

Is this a valid nit? Basically if we do storage read and then origin check, `none` origin could spam a storage read?

**Review by shawntabrizi — APPROVED** [review]

**Pull request merged**